### ISSUE TITLE
Unannotated decorator that replaces a class with a dynamically generated subclass is not understood

### ISSUE THREAD

**Issue by ghimiresdp** [astral-sh/ty#2604]
### Summary

Whenever we have a decorated class, `ty` does not resolve attributes added in the decorator.

In the code below, there is a `personify` decorator that has following attributes added:

- full_name: `str`
- set_full_name: `callable`

### example code:

```py
def personify(cls):
    class Wrapped(cls):
        full_name: str

        def set_full_name(self, full_name: str):
            self.full_name = full_name

    return Wrapped


@personify
class Animal:
    pass


animal = Animal()
animal.set_full_name("John")
print(f"The full name is: {animal.full_name}")

```
### `uvx ty check` output
<img width="1055" height="397" alt="Image" src="https://github.com/user-attachments/assets/275dc216-bd7b-4728-b579-8569654f7f6a" />

### Editor Output
<img width="1319" height="769" alt="Image" src="https://github.com/user-attachments/assets/17579d80-ffb3-49c4-870d-9361cbed3819" />

### Version

0.0.13

**Comment by AlexWaygood** [comment]
Thanks for the report!

I assumed that no other type checker would support this, because:
- Your `personify` decorator is unannotated, so there's nothing to indicate to the type checker that it returns a type different to the one that was passed in; and,
- Even if it were annotated, it would be hard (if not impossible) to annotate this decorator in a precise way

In an ideal world, maybe you could annotate your decorator with something like this, but even that won't work until https://github.com/astral-sh/ty/issues/143 is implemented, and it has to use nonstandard extensions from our `ty_extensions` module (which doesn't exist at runtime):

```py
from typing import Protocol, TYPE_CHECKING

if TYPE_CHECKING:
    from ty_extensions import Intersection


class WrappedLike(Protocol):
    full_name: str

    def set_full_name(self, full_name: str) -> None: ...


def personify[T: type](cls: T) -> Intersection[T, type[WrappedLike]]:
    class Wrapped(cls):
        full_name: str

        def set_full_name(self, full_name: str):
            self.full_name = full_name

    return Wrapped


@personify
class Animal:
    pass


animal = Animal()
animal.set_full_name("John")
print(f"The full name is: {animal.full_name}")
```

As I expected, [mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=79378147bc02bdd0a5c08958b9f8e4a9) and [pyrefly](https://pyrefly.org/sandbox/?project=N4IgZglgNgpgziAXKOBDAdgEwEYHsAeAdAA4CeSImMYABMTAE5y7oRikAUAxlHAJSIAOuhqiaPVHDg0A6g1TF6mbrwHCxGmmACuUKAH10qALYxENOABcGw9ZtFVacGJf069hkzA7OoYADRaugZGpuZWDGoi9pq%2BYITuIV40ALxBHqEwttGiDC7aDCJyCkrZwgAC9EwsbKTCElI0AIKsxqhQQjl0knBlRq3tqc0DUBx8whgQbVCEzq6JnqYcgiAAUrgAFugr4%2BjEDBDolhxgKwAqGzDpUDSZNBBw5sCT0wnBizAAvjvCIP4g2ks0DgJHIiBAAGIaABVIFQCCWUhBdBcIEsXrbLDULS4BhtVzobTGbCMDj4cyHSx8GgAWgAfBZrJ0NHlLAURKcQAA5IkkhhPfDfEC-f5kPJgKCkQiWXDGKAUKEABVI4slFgwOAI4hYkAA5gVUGj0IRhFCAMowK4bSyWYiPAD09rF1ElhFxuvtMHQ9swuC4cHtXB1EH18iN9pxDBoqAAbqhoKhsLBteg9QajTRcMQjSDhGRLBsWDSY4w4BAWEMVgBmQgARgATCtfp9-qhURASwAxaAwChoLB4IhkECfIA) do not support this decorator, therefore. But to my surprise, [pyright does](https://pyright-play.net/?pyrightVersion=1.1.405&pythonVersion=3.13&reportUnreachable=true&code=CYUwZgBADiBODOB7AdgSzATwBQGMA28AlAFwBQEFE%2BAhvPBAOqzVQzC4EnmU9gCuePAH1k1ALYhiEeABdYpbjwqhI8EDKH9BI8SCxq8YADQQtw0RKmzYXJXekhDAOjM6JEALymB53Qp6w6nywyIzMrCDACqQAAjAIKOgYpDR0EACCaGLUeGQ8ULTw0dRZOZ4ZpXhYhKQlqNl4TmoarhZ6AEQAUogAFsjtNVCwqMgyWGDtACo9IN6CEG0QqPBSAN51DS4%2BbiAAvgOkQA)! It looks like it _may_ be inlining the entire contents of the decorator at the callsite, which is really impressive.

I'd love for ty to be able to match pyright's sophistication there at some point, but it probably won't happen in the near future, I'm afraid! This requires some pretty advanced features from us.

**Comment by hauntsaninja** [comment]
You're giving pyright ~too much credit~ deserved credit although its behaviour is not perfect: https://pyright-play.net/?pyrightVersion=1.1.405&pythonVersion=3.13&reportUnreachable=true&code=CYUwZgBADiBODOB7AdgSzATwBQGMA28AlAFwBQEFE%2BAhvPBAOqzVQzC4EnmU9gCuePAH1k1ALYhiEeABdYpbjwqhI8EDKH9BI8SCxq8YADQQtw0RKmzYXJXekhDAOjM6JEALymB53Qp6w6nywyIzMrCDACqQAAjAIKOgYpDR0EACCaGLUeGQ8ULTw0dRZOZ4ZpXhYhKQlqNl4TmoarhZ6AEQAUogAFsjtNVCwqMgyWGDtACo9IN6CEG0QqPBSAN51DS4%2BbiBCACIA8gCiAMpCAHIHk0JHABoAkieTAL4DpEA

Edit: hmm although it does reveal_type as Wrapped, so I think it is doing some inlining but treating Wrapped as having an unknown base class

**Comment by AlexWaygood** [comment]
Yeah I chatted with @carljm about this earlier — looks like it's doing return type inference here but not call inlining (but it does do inlining for smaller unannotated functions in some situations). Either way, what pyright is doing is very cool, and either way, it's still an advanced feature that's blocked on multiple things for us right now 😄

**Comment by carljm** [comment]
It's worth noting also that return-type inference is generally unsound with returned classes, since the type checker will assume there is one copy of that class, not N copies, and this can easily lead to wrong conclusions with e.g. `isinstance` checks. If you have two different classes decorated with this `@personify` decorator, pyright will think they are exactly the same class, and draw [incorrect reachability conclusions](https://pyright-play.net/?pyrightVersion=1.1.405&pythonVersion=3.13&reportUnreachable=true&code=CYUwZgBADiBODOB7AdgSzATwBQGMA28AlAFwBQEFE%2BAhvPBAOqzVQzC4EnmU9gCuePAH1k1ALYhiEeABdYpbjwqhI8EDKH9BI8SCxq8YADQQtw0RKmzYXJXekhDAOjM6JEALymB53Qp6w6nywyIzMrCDACqQAAjAIKOgYpDR0EACCaGLUeGQ8ULTwCnFwSGiYKXiFEAAqABaoyADmeZQFdNEqplgAHlKZqNl4tpToEMiIMhCo8I2y1Mg4ej0m9Y1NI3YAjBAA1BAARGCIiAeKEKRAA) accordingly.

So I don't think we will want to emulate pyright's behavior here -- our intention is to exclude classes from RTI. I think the better way to support this in ty would be to annotate the decorator as returning an intersection of the original class and a callable type returning a protocol representing the added interface. (The intersection part of this works already today -- using it as a class decorator requires #143). If the decorator is un-annotated, we will likely just assume it doesn't change the class (the alternative would be to consider the class entirely `Unknown`.)

Given that, I'm going to remove the "blocking" relationship with RTI here.
